File: csharp-sdk/src/Cleanster/Api/PaymentMethodsApi.cs
using System.Text.Json;
using Cleanster.Models;

namespace Cleanster.Api;

/// <summary>Manages Stripe and PayPal payment methods.</summary>
public sealed class PaymentMethodsApi
{
    private readonly ICleansterHttpClient _http;
    internal PaymentMethodsApi(ICleansterHttpClient http) => _http = http;

    /// <summary>
    /// Return Stripe SetupIntent details for client-side card collection.
    /// Use the returned <c>clientSecret</c> with Stripe.js on the client.
    /// </summary>
    public async Task<ApiResponse<JsonElement>> GetSetupIntentDetailsAsync(CancellationToken ct = default)
    {
        var root = await _http.GetAsync("/v1/payment-methods/setup-intent-details", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Return a PayPal client token for rendering the PayPal button client-side.</summary>
    public async Task<ApiResponse<JsonElement>> GetPaypalClientTokenAsync(CancellationToken ct = default)
    {
        var root = await _http.GetAsync("/v1/payment-methods/paypal-client-token", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Save a new payment method after client-side tokenization.</summary>
    /// <param name="paymentMethodId">The tokenized payment method ID from Stripe.js or PayPal.</param>
    public async Task<ApiResponse<JsonElement>> AddPaymentMethodAsync(
        string paymentMethodId, CancellationToken ct = default)
    {
        var root = await _http.PostAsync("/v1/payment-methods", new { paymentMethodId }, ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Return all saved payment methods for the current user.</summary>
    public async Task<ApiResponse<List<PaymentMethod>>> GetPaymentMethodsAsync(CancellationToken ct = default)
    {
        var root = await _http.GetAsync("/v1/payment-methods", ct: ct);
        return JsonHelper.ParseList<PaymentMethod>(root);
    }

    /// <summary>Remove a saved payment method.</summary>
    public async Task<ApiResponse<JsonElement>> DeletePaymentMethodAsync(
        int paymentMethodId, CancellationToken ct = default)
    {
        var root = await _http.DeleteAsync($"/v1/payment-methods/{paymentMethodId}", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Make a payment method the default for future bookings.</summary>
    public async Task<ApiResponse<JsonElement>> SetDefaultPaymentMethodAsync(
        int paymentMethodId, CancellationToken ct = default)
    {
        var root = await _http.PutAsync($"/v1/payment-methods/{paymentMethodId}/default", ct: ct);
        return JsonHelper.ParseRaw(root);
    }
}

File: android-sdk/src/main/kotlin/com/cleanster/android/api/PaymentMethodsApi.kt
package com.cleanster.android.api

import com.cleanster.android.model.*
import retrofit2.Retrofit
import retrofit2.http.*

internal interface PaymentMethodsService {
    @GET("v1/payment-methods/setup-intent-details")
    suspend fun getSetupIntentDetails(): ApiResponse<Any>

    @GET("v1/payment-methods/paypal-client-token")
    suspend fun getPayPalClientToken(): ApiResponse<Any>

    @POST("v1/payment-methods")
    suspend fun addPaymentMethod(@Body body: AddPaymentMethodRequest): ApiResponse<PaymentMethod>

    @GET("v1/payment-methods")
    suspend fun getPaymentMethods(): ApiResponse<List<PaymentMethod>>

    @PUT("v1/payment-methods/{paymentMethodId}/default")
    suspend fun setDefaultPaymentMethod(
        @Path("paymentMethodId") paymentMethodId: Int,
    ): ApiResponse<Any>

    @DELETE("v1/payment-methods/{paymentMethodId}")
    suspend fun deletePaymentMethod(
        @Path("paymentMethodId") paymentMethodId: Int,
    ): ApiResponse<Any>
}

class PaymentMethodsApi(retrofit: Retrofit) {
    private val service = retrofit.create(PaymentMethodsService::class.java)

    suspend fun getSetupIntentDetails() = wrap { service.getSetupIntentDetails() }

    suspend fun getPayPalClientToken() = wrap { service.getPayPalClientToken() }

    suspend fun addPaymentMethod(paymentMethodId: String) =
        wrap { service.addPaymentMethod(AddPaymentMethodRequest(paymentMethodId)) }

    suspend fun getPaymentMethods() = wrap { service.getPaymentMethods() }

    suspend fun setDefaultPaymentMethod(paymentMethodId: Int) =
        wrap { service.setDefaultPaymentMethod(paymentMethodId) }

    suspend fun deletePaymentMethod(paymentMethodId: Int) =
        wrap { service.deletePaymentMethod(paymentMethodId) }
}
